src/net/stream.rs

```diff
@@ -1,4 +1,188 @@
+            return Poll::Ready(Ok(()));
+        }
+
+        match ready!(self.receiver.poll_recv(cx)) {
+            Some(data) => {
+                trace!(dst = ?self.pair.local, src = ?self.pair.remote, protocol = %data, "Recv");
+
+                let bytes = data.0.as_ref();
+                buf.put_slice(bytes);
+                self.is_closed = bytes.is_empty();
+
+                Poll::Ready(Ok(()))
+            }
+            None => Poll::Ready(Err(io::Error::new(
+                io::ErrorKind::ConnectionReset,
+                "Connection reset",
+            ))),
+        }
+    }
+
+    fn poll_write_priv(&self, _cx: &mut Context<'_>, buf: &[u8]) -> Poll<Result<usize>> {
+        if buf.remaining() == 0 {
+            return Poll::Ready(Ok(0));
+        }
+
+        if self.is_shutdown {
+            return Poll::Ready(Err(io::Error::new(
+                io::ErrorKind::BrokenPipe,
+                "Broken pipe",
+            )));
+        }
+
+        let res = World::current(|world| {
+            let bytes = Bytes::copy_from_slice(buf);
+            let len = bytes.len();
+
+            let seq = self.seq(world)?;
+            self.send(world, Segment::Data(seq, StreamData(bytes)));
+
+            Ok(len)
+        });
+
+        Poll::Ready(res)
+    }
+
+    fn poll_shutdown_priv(&mut self) -> Poll<Result<()>> {
+        if self.is_shutdown {
+            return Poll::Ready(Err(io::Error::new(
+                io::ErrorKind::NotConnected,
+                "Socket is not connected",
+            )));
+        }
+
+        let res = World::current(|world| {
+            let seq = self.seq(world)?;
+            self.send(world, Segment::Fin(seq));
+
+            self.is_shutdown = true;
+
+            Ok(())
+        });
+
+        Poll::Ready(res)
+    }
+
+    // If a seq is not assignable the connection has been reset by the
+    // peer.
+    fn seq(&self, world: &mut World) -> Result<u64> {
+        world
+            .current_host_mut()
+            .tcp
+            .assing_send_seq(self.pair)
+            .ok_or_else(|| io::Error::new(io::ErrorKind::BrokenPipe, "Broken pipe"))
+    }
+
+    fn send(&self, world: &mut World, segment: Segment) {
+        world.send_message(self.pair.local, self.pair.remote, Protocol::Tcp(segment));
+    }
+}
+
+impl AsyncRead for TcpStream {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf,
+    ) -> Poll<Result<()>> {
+        self.poll_read_priv(cx, buf)
+    }
+}
+
+impl AsyncWrite for TcpStream {
+    fn poll_write(self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &[u8]) -> Poll<Result<usize>> {
+        self.poll_write_priv(cx, buf)
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_shutdown(mut self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<()>> {
+        self.poll_shutdown_priv()
+    }
+}
```

Comment:
i think adding spans to each of these impls would be quite good useful, i think. you'd probably want to have a single persistent span corresponding to `TcpStream` that gets entered on each method call.